### PR TITLE
feat(pacmak): prerelease identifier support

### LINKED ISSUE #2114: [pacmak] support pre-release versions
### Affected Languages
- [x] `TypeScript` or `Javascript`
- [x] `Python`
- [x] `Java`
- [x] .NET (`C#`, `F#`, ...)

### Description

We would like to release the the next major version of `aws-cdk` and the `constructs` libraries
initially as pre-release. See [cdk pre-release identifiers](https://github.com/aws/aws-cdk-rfcs/blob/master/text/0079-cdk-2.0.md#releasing-to-package-managers) and [`constructs` pre-release identifier](https://github.com/aws/constructs/commit/c9b0726008dd9b5148c6efea5341a0f1c1f62715#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R126).

JSII packing needs to natively support this.

### Proposed Solution

If the `version` field in `package.json` contains a pre-release suffix, JSII should treat the suffix as a pre-release identifier and create pre-release versions appropriate for the package managers of the target languages.

For example, if `version` is specified as `2.0.0-alpha.1`, it must automatically detect that this is a pre-release version and create target versions as `2.0.0-alpha.1` for Maven Central, `2.0.0-alpha.1` for Nuget and `2.0.0.a1` for PyPI.

Some package managers are restrictive in what is acceptable pre-release versions. For example, PyPI [only allows](https://packaging.python.org/guides/distributing-packages-using-setuptools/#pre-release-versioning) `devN`, `aN`, `bN` and `rcN` (where `N` is the pre-release version number) as valid pre-release versions. 
For this reason, when enabled for these package managers, it must make reasonable assumptions about allowed pre-release identifiers, erring if a pre-release identifier cannot be translated and allow overriding them in the `jsii` section.

### PR DESCRIPTION
**Pull request by RomainMuller**
When an NPM package built with `jsii` is tagged with a version that
includes a prerelease identifier (e.g: `1.2.3-alpha.1`), automatically
represent this as a pre-release version in generated packages, too. For
most currently supported languages, this simply means forwarding the
version untouched, however for Python, prerelease identifiers need to
be mapped into one of the supported tokens in PyPI (`.dev#`, `.a#`,
`.b#` or `.rc#`).

Fixes #2114

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/version-utils.ts:77]:
```suggestion
 * @param target          the target language for which the version is destined
```

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/version-utils.ts:129]: throw an error here

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/version-utils.ts:142]: throw an error here as well.

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/version-utils.ts:232]: The existing implementation looks as though one form of version string (semver) is being converted to another form (bracket form?). 
Perhaps leave this translation 'pure' and move the call to `toReleaseVersion()` one level above?

File: packages/jsii-pacmak/lib/targets/dotnet.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/dotnet.ts:278]: Something is off here in terms of code cleanliness. 

Ideally, besides the jsii `Assembly` context, there's also a pacmak specific context that gets passed around, which converts `assembly.version` into the release version identifier. So the conversion of one version to the other happens only once. Something like -

```ts
interface PacMakContext {
  public readonly assm: spec.Assembly;
  // ...
  public readonly releaseVersion;
}
```

This would "home" the translation in the correct place.

Do you see value in something like that?

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/version-utils.ts:95]: Hmmm. Do we not have an enum or something that represents the list of supported target langs?

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/version-utils.ts:123]: We should allow user overrides of the pre-release suffix, per language, in the `package.json`? 

Useful specifically for target package managers, like Python, that don't allow *any* pre-release identifier.
*use case:* customer releases pre-release version of their library in 1 language and now wants to expand to all, but has used a pre-release identifier that doesn't automatically map.

Comment (nija-at) [reply packages/jsii-pacmak/lib/targets/version-utils.ts:123]: Another reason this might be worth adding is that it's not clear what Python allows or disallows. [This spec](https://packaging.python.org/guides/distributing-packages-using-setuptools/#pre-release-versioning) states "may be published as" which is ambiguous.

And [PEP-440](https://www.python.org/dev/peps/pep-0440/#pre-release-spelling) states 'alpha' and 'beta' are allowed and, 'pre' represents release candidate. Is PEP-440 active?

Comment (nija-at) [reply packages/jsii-pacmak/lib/targets/version-utils.ts:123]: Might have found the first use case 😅

https://github.com/aws/constructs/blob/2a6bb57f02a22d844b71b97428f2545afb4f87ea/package.json#L126

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/version-utils.ts:232]: The bracket form is a *range* transformation... And the range has versions on both sides, which need correct handling.

File: packages/jsii-pacmak/lib/targets/dotnet.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/dotnet.ts:278]: Could be. I mean this is also why long term I'm interested in having code-gen happen within `jsii` directly...

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/version-utils.ts:95]: Not an enum; but the `TargetName` is a string-literal-union type, so one can't make typos here without the compiler noticing.

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/version-utils.ts:123]: I would map `pre` to `dev` and the `constructs` problem would be gone?

I'm ambivalent about adding more knobs to override behaviors in the configuration files; because the onus is then on the user to know what is acceptable or not (so now, they have to go look at PEP-440, and get confused by how it seems to say something different from `pip`'s documentation....). I would rather add such a feature only once it has been demonstrated to be the only way to unblock a reasonable use-case.

File: packages/jsii-pacmak/lib/targets/dotnet.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/dotnet.ts]: Also - the problem with your suggestion is that it would need to carry a context... for *each* dependency, too. Because we also need to perform the prerelease mapping for dependency ranges (otherwise we'll have resolution issues)

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/version-utils.ts:81]: We need this entire pre-release support in pacmak and the python quirks documented somewhere.

I see this package doesn't have a proper README.
Can we start by adding this?

Comment (nija-at) [reply packages/jsii-pacmak/lib/targets/version-utils.ts:81]: As a side note, I recommend adding something similar to 'prlinter' for jsii, so that all `feat` require a README update. Pretty useful IMO.

File: packages/jsii-pacmak/lib/targets/dotnet.ts
Comment (nija-at) [reply packages/jsii-pacmak/lib/targets/dotnet.ts]: I'll let you decide how useful or not it is. I've barely worked in this side of the code base.
Just feels like something we could centralize `toReleaseVersion()` and make it available via a context.

File: packages/jsii-pacmak/test/targets/version-utils.test.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/test/targets/version-utils.test.ts:123]: Only `-pre` is not supported nor supportable.
Python *requires* a sequence identifier (so `-pre.#`).

If I make `-pre` map to `.dev0`, then how do I distinguish it form the (also valid) `-pre.0`?

File: packages/jsii-pacmak/test/targets/version-utils.test.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/test/targets/version-utils.test.ts:150]: This is, in fact, guaranteed by the TypeScript compiler, due to how expectations are defined as:

```ts
type Expectations = { readonly [K in TargetName]: string | RegExp };
```

This makes valid values require to have keys in `TargetName` (and nothing else), with none of them optional (so all of them required). I verified this while authoring the PR (I was pleasantly surprised)

File: packages/jsii-pacmak/test/targets/version-utils.test.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/test/targets/version-utils.test.ts:107]: That is precisely why this PR is still a *Draft*. I don't quite have a good way to test for this right now... And I have to come up with something.

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/version-utils.ts:81]: > As a side note, I recommend adding something similar to 'prlinter' that we have in the cdk for the jsii repo. Enforce that all feat require a README update. Pretty useful IMO.

I have mixed feelings on this based on how often I have needed to apply labels to bypass this validation in my CDK contributions (for good reasons) in the past. But you're probably right, still... At least it ensures people have fought about it.

Anyhow - the fact there is no README.md entry yet is because this is a draft and I didn't want to have to come back to it multiple times if the implementation changes dramatically as I get into testing it properly.

File: docs/configuration.md
Comment (nija-at) [inline docs/configuration.md:4]: (for the future) prefer wholesale adjustments and refactors to be separate PRs, hence separate commits 😉

File: packages/jsii-pacmak/lib/targets/dotnet/filegenerator.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/dotnet/filegenerator.ts:187]: Non-blocking: It's a consolation that in most languages (besides .NET), `toReleaseVersion()` is called once or from a single file. Is it simple to do the same with .NET?

File: packages/jsii-pacmak/lib/targets/js.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/js.ts:17]: Call `toReleaseVersion()` once in this code block.

```ts
const releaseVersion = toReleaseVersion(assm.version, TargetName.JAVASCRIPT);
```

File: packages/jsii-pacmak/lib/targets/js.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/js.ts:22]: just checking - need to call `toReleaseVersion()` here?

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (nija-at) [reply packages/jsii-pacmak/lib/targets/version-utils.ts:81]: I understand it's annoying to add the label, but the value outweighs it, IMO.

File: packages/jsii-pacmak/lib/targets/version-utils.ts
Comment (nija-at) [inline packages/jsii-pacmak/lib/targets/version-utils.ts:114]: Link to the pep spec would be useful here

File: docs/configuration.md
Comment (nija-at) [inline docs/configuration.md:201]: I don't see a use case for why we should support post releases or development versions of post releases. They should be new patch/minor versions, or new pre-release identifiers.

I can only find this in Python and even PEP-440 discourages them -

> The use of post-releases to publish maintenance releases containing actual bug fixes is strongly discouraged. In general, it is better to use a longer release number and increment the final component for each maintenance release.

> Creating post-releases of pre-releases is strongly discouraged, as it makes the version identifier difficult to parse for human readers. In general, it is substantially clearer to simply create a new pre-release by incrementing the numeric component.

**Review by nija-at — CHANGES_REQUESTED** [review]
My main concern is about post releases.

File: docs/configuration.md
Comment (RomainMuller) [reply docs/configuration.md:4]: Yeah this is auto-format-on-save... -_-

File: packages/jsii-pacmak/lib/targets/dotnet/filegenerator.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/dotnet/filegenerator.ts:187]: Maybe it would be but this could entail a significant refactor... And perhaps would reduce the overall quality of the .NET generator's code quality 🙃

File: packages/jsii-pacmak/lib/targets/js.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/js.ts:22]: I guess yes (for consistency). But `toReleaseVersion` is basically no-op for JavaScript.

File: docs/configuration.md
Comment (RomainMuller) [reply docs/configuration.md:201]: I mean I don't intend on those post-releases being published at all. I conventionally use `-pr<pr-id>.0` version suffix for PR-bound pre-releases (i.e: the artifacts that would be produced if that PR was to be merged).

So I found it actually mapped kind of neatly (this is the post-release of this thing for pr `<pr-id>`, sequence `0`).

Would it be "better" to you if I'd simply map `X.Y.Z-prN.0` to `X.Y.Z.devN` (and fail if the sequence is non-0)?

Comment (nija-at) [reply docs/configuration.md]: > Would it be "better" to you if I'd simply map X.Y.Z-prN.0 to X.Y.Z.devN (and fail if the sequence is non-0)?

Better yes, but still doesn't work. 

Do we have a use case for PR-bound pre-releases? What's the value over using `dev` prefix with an auto-inc sequence instead?

File: packages/jsii-pacmak/test/targets/version-utils.test.ts
Comment (nija-at) [reply packages/jsii-pacmak/test/targets/version-utils.test.ts:123]: Makes sense. I mis-read.

File: packages/jsii-pacmak/lib/targets/dotnet/filegenerator.ts
Comment (nija-at) [reply packages/jsii-pacmak/lib/targets/dotnet/filegenerator.ts:187]: Alright-y. Not ideal but ok :)

File: docs/configuration.md
Comment (nija-at) [inline docs/configuration.md:236]:
```suggestion
[maven central]: https://search.maven.org
```

**Review by nija-at — APPROVED** [review]

File: docs/configuration.md
Comment (RomainMuller) [reply docs/configuration.md]: Alright - dropping support for this, will re-instate if actually needed.

**Pull request merged**
